File: Java/1268.search-suggestions-system.java
/*
 * @lc app=leetcode id=1268 lang=java
 *
 * [1268] Search Suggestions System
 */

// @lc code=start
/*
先将products排序，然后建Trie,每个node存对应prefix的product(最多3个)
time: O(nlogn + O(n * len(product) for building trie + O(len(searchWord) for query)
space: O(n * len(product))
* 注意query的写法, leetcode有一个很复杂的corner case
*/
class Solution {
    class TrieNode {
        List<String> list;
        TrieNode[] children;
        public TrieNode() {
            list = new ArrayList<>();
            children = new TrieNode[26];
        }
    }
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        List<List<String>> res = new ArrayList<>();
        Arrays.sort(products);
        TrieNode root = new TrieNode();
        TrieNode cur = root;
        for (String product : products) {
                cur = root;
            for (char c : product.toCharArray()) {
                if (cur.children[c - 'a'] == null) cur.children[c - 'a'] = new TrieNode();
                cur = cur.children[c - 'a'];
                if (cur.list.size() < 3) cur.list.add(product);
            }
        }
        cur = root;
        for (char c : searchWord.toCharArray()) {
            if (cur != null) cur = cur.children[c - 'a'];
            res.add(cur == null ? new ArrayList<>() : cur.list);
        }
        return res;
    }
}
// @lc code=end

File: src/1268.search-suggestions-system.cpp
/*
 * @lc app=leetcode id=1268 lang=cpp
 *
 * [1268] Search Suggestions System
 */

// @lc code=start
/*
binary search,找prefix的lower bound + upper bound
upper bound用"key + ~"来代替(~的ascii > z)
time: sortO(nlogn) queryO(length of words * logn)
space: O(1)
*/
class Solution {
public:
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        vector<vector<string>>res;
        sort(products.begin(), products.end());
        string prefix = "";
        for (char c : searchWord) {
            prefix += c;
            auto low = lower_bound(products.begin(), products.end(), prefix);
            auto high = upper_bound(products.begin(), products.end(), prefix + '~');
            if (low == high) break;
            res.emplace_back(low, min(low + 3, high));
            // res.push_back(vector<string>(low, min(low + 3, high)));
        }
        while (res.size() < searchWord.size()) res.push_back({});
        return res;
    }
};
// @lc code=end

